README.md
# mapper-openc

## Overview

The Open Corporates mapper python scripts convert the Open Corporates company and officers located
[here](https://opencorporates.com/info/our-data//).  While you can search individual companies for free, you will need to purchase 
the regions you desire for loading into Senzing. You can purchase and download files of companies, their officers, and their additional 
(non-registered) addresses.

There are two python scripts in this project ...
- The [openc-load-childb.py](openc-load-childb.py) script creates the child file database for additional addresses, aliases and identifiers.
- The [openc-companies.py](openc-companies.py) script maps the companies with their additional addresses if present. 
- The [openc-officers.py](openc-officers.py) script just maps the officers and relates them to their company.

Loading this data into Senzing requires additional features and configurations. These are contained in the
[openc-config-updates.g2c](openc-config-updates.g2c) file.

## Contents

1. [Prerequisites](#prerequisites)
2. [Download Open Corporates files](#download-open-corporates-files)
3. [Installation](#installation)
4. [Configuring Senzing](#configuring-senzing)
5. [Create the child database](#create-the-child-database)
6. [Running the companies mapper](#running-the-companies-mapper)
7. [Running the officers mapper](#running-the-officers-mapper)

### Prerequisites

- python 3.6 or higher
- Senzing API version 2.1 or higher
- pandas (pip3 install pandas)
- orjson (pip3 install orjson)

### Download Open Corporates files

When you purchase Open Corporates bulk data, you will be given instructions for how to download their files.  These are the expected files ...

These are the primary files:
- **companies.csv.gz** This is the list of companies.
- **officers.csv.gz** This is the list of officers for the companies.

These are child files containing additional information for companies:
- **alternative_names.csv.gz** This is a list of additional names for the companies.
- **non_reg_addresses.csv.gz** This is a list of additional addresses for the companies.
- **additional_identifiers.csv.gz** This is a list of tax and other identifiers for the companies.

Place all these files on a single directory refered to as the "input" directory below.

### Installation

Place the the following files on a directory of your choice ...

- [openc-config-updates.g2c](openc-config-updates.g2c)
- [openc-load-childb.py](openc-load-childb.py)
- [openc-companies.py](openc-companies.py)
- [openc-officers.py](openc-officers.py)

### Configuring Senzing

*Note:* This only needs to be performed one time! In fact you may want to add these configuration updates to a master configuration file for all your data sources.

From your Senzing project directory ...

```console
python3 G2ConfigTool.py <path-to-file>/openc_config_updates.g2c -f
```

This will step you through the process of adding the data sources, features, attributes and other settings needed to load this data into 
Senzing. After each command you will see a status message saying "success" or "already exists". For instance, if you run the script twice, 
the second time through they will all say "already exists" which is OK.


### Create the child database

*This step must be performed before running the mappers!*

```console
python3 openc-load-childb.py --help
usage: openc-load-childb.py [-h] [-i INPUT_FILE_DIR] [-c CHILD_DATABASE_NAME]

optional arguments:
  -h, --help            show this help message and exit
  -i INPUT_FILE_DIR, --input_file_dir INPUT_FILE_DIR
                        the name of the open corporates csv file directory
  -c CHILD_DATABASE_NAME, --child_database_name CHILD_DATABASE_NAME
                        the name of the database file to create
```

Typical use:
```console
python3 openc-load-childb.py -i ./input -c ./input/child.db
```

- The -i should be the directory where you downloaded the Open Corporates data files.
- The -c should be where you want the sqlite child database to be written.  Ideally, you would place it on the same directory.


### Running the companies mapper

```console
python3 openc-companies.py --help
usage: openc-companies.py [-h] [-i INPUT_FILE_DIR] [-o OUTPUT_FILE_DIR] [-c CHILD_DATABASE_NAME] [-d DATA_SOURCE] [-l LOG_FILE]

optional arguments:
  -h, --help            show this help message and exit
  -i INPUT_FILE_NAME, --input_file_name INPUT_FILE_NAME
                        the name of an open corporates csv file for companies
  -o OUTPUT_FILE_NAME, --output_file_name OUTPUT_FILE_NAME
                        the name of the output file
  -c CHILD_DATABASE_NAME, --child_database_name CHILD_DATABASE_NAME
                        the name of the child database created in the prior step
  -d DATA_SOURCE, --data_source DATA_SOURCE
                        the name of the data source code to use, defaults to: OPENC-COMPANY
  -l LOG_FILE, --log_file LOG_FILE
                        optional name of the statistics log file
  -w MAX_WORKERS, --max_workers MAX_WORKERS
                        defaults to the number of system processors, may need to reduce if running other things at same time
```

Typical use: 
```console
python3 openc-companies.py -i ./input/companies.csv.gz -c ./input/child.db -o ./output/companies.json -l output/company-mapping-log.json
```

- The -i is location of the Open Corporates companies data file.
- The -c is where the child database you created in the prior step is located.
- The -o is where you want the mapped file to be written.
- The -l is an optional log file that contains mapping stats for your review.


### Running the officers mapper

```console
python3 openc-officers.py --help
usage: openc-officers.py [-h] [-i INPUT_FILE_SET] [-o OUTPUT_FILE_DIR] [-d DATA_SOURCE] [-l LOG_FILE]

optional arguments:
  -h, --help            show this help message and exit
  -i INPUT_FILE_NAME, --input_file_name INPUT_FILE_NAME
                        the name of an open corporates csv file for officers
  -o OUTPUT_FILE_NAME, --output_file_name OUTPUT_FILE_NAME
                        the name of the output file
  -t TEMP_DATABASE_NAME, --temp_database_name TEMP_DATABASE_NAME
                        the name for the temporary database required to de-dupe officers
  -d DATA_SOURCE, --data_source DATA_SOURCE
                        the name of the data source code to use, defaults to: OPENC-OFFICER
  -l LOG_FILE, --log_file LOG_FILE
                        optional name of the statistics log file
```

Typical use: 
```console
python3 openc-officers.py -i ./input/officers.csv.gz -o ./output/officers.json -t ./input/temp.db -l output/officers-log.json
```

- The -i is the location of the Open Corporates officers data file.
- The -o is where you want the mapped file to be written.
- The -t is for the temporary sqlite database used to de-dupe officers.
- The -l is an optional log file that contains mapping stats for your review.

*Note:* The temporary database file will be overwritten if exists! You can delete it manually after the run to save disk space.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
[markdownlint](https://dlaa.me/markdownlint/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased] - 2023-01-04

- Overall performance increases
- Added officer compression of pure duplicates


IOQueueProcessor.py
try:
    import multiprocess as multiprocessing
except:
    import multiprocessing

from queue import Empty, Full
import signal
import time

class IOQueueProcessor():

    def __init__(self, input_class, output_class, **kwargs):

        self.process_count = kwargs.get('process_count', multiprocessing.cpu_count())
        self.all_stop = multiprocessing.Value('i', 0)

        self.input_class = input_class
        self.output_class = output_class
        self.input_queue = multiprocessing.Queue(self.process_count * 10)
        self.output_queue = multiprocessing.Queue(self.process_count * 10)

        self.input_queue_read_cnt = multiprocessing.Value('i', 0)
        self.output_queue_read_cnt = multiprocessing.Value('i', 0)

        self.kwargs = kwargs
        self.process_list = []

    def start_up(self):

        self.process_list.append(multiprocessing.Process(target=self.output_queue_reader, args=(0, self.output_queue, self.output_class), kwargs=self.kwargs))
        for process_number in range(self.process_count - 1):
            self.process_list.append(multiprocessing.Process(target=self.input_queue_reader, args=(process_number+1, self.input_queue, self.output_queue, self.input_class), kwargs=self.kwargs))
        for process in self.process_list:
            process.start()

    def finish_up(self):

        # wait for queues
        try:
            while self.input_queue.qsize() or self.output_queue.qsize():
                print(f"waiting for {self.input_queue.qsize()} input and {self.output_queue.qsize()} output queue records")
                time.sleep(1)
        except: # qsize does not work on mac
            while not self.input_queue.empty() or not self.output_queue.empty():
                if not self.input_queue.empty():
                    print("waiting for input queue to finish")
                elif not self.output_queue.empty():
                    print("waiting for output queue to finish")
                time.sleep(1)

        with self.all_stop.get_lock():
            self.all_stop.value = 1

        start = time.time()
        while time.time() - start <= 15:
            if not any(process.is_alive() for process in self.process_list):
                break
            time.sleep(1)

        for process in self.process_list:
            if process.is_alive():
                print(process.name, 'did not terminate gracefully')
                process.terminate()
            process.join()

        self.input_queue.close()
        self.output_queue.close()

    def queue_read(self, q):
        try:
            return q.get(True, 1)
        except Empty:
            return None

    def queue_write(self, q, msg):
        while True:
            try:
                q.put(msg, True, 1)
            except Full:
                continue
            break

    def input_queue_reader(self, process_number, input_queue, output_queue, function_ref, **kwargs):

        kwargs['process_number'] = process_number
        input_class = function_ref(**kwargs)

        while self.all_stop.value == 0:
            queue_data = self.queue_read(input_queue)
            if queue_data:
                with self.input_queue_read_cnt.get_lock():
                    self.input_queue_read_cnt.value += 1
                result = input_class.run(queue_data)
                if result:
                    self.queue_write(output_queue, result)

        input_class.close()

    def output_queue_reader(self, process_number, output_queue, function_ref, **kwargs):

        kwargs['process_number'] = process_number
        output_class = function_ref(**kwargs)

        while self.all_stop.value == 0:
            queue_data = self.queue_read(output_queue)
            if queue_data:
                with self.output_queue_read_cnt.get_lock():
                    self.output_queue_read_cnt.value += 1
                output_class.run(queue_data)

        output_class.close()

    def process(self, msg):
        self.queue_write(self.input_queue, msg)

    def get_input_queue_read_cnt(self):
        return self.input_queue_read_cnt.value

    def get_output_queue_read_cnt(self):
        return self.output_queue_read_cnt.value


class reader():

    def __init__(self, **kwargs):
        self.process_number = kwargs.get('process_number', -1)
        self.dbo = kwargs['dbo']
        print(f"process {self.process_number} opened {self.dbo}")

    def close(self):
        print(f"process {self.process_number} closed {self.dbo}")

    def run(self, raw_data):
        return f"process {self.process_number} returned {raw_data} mapped!"

class writer():

    def __init__(self, **kwargs):
        self.process_number = kwargs.get('process_number', -1)
        self.output_file_name = kwargs['output_file_name']
        print(f"process {self.process_number} opened {self.output_file_name}")
        self.output_data = []

    def close(self):
        print(f"process {self.process_number} closed {self.output_file_name}")

    def run(self, mapped_data):
        self.output_data.append(mapped_data)
        return f"process {self.process_number} wrote {mapped_data}"


def signal_handler(signal, frame):
    with shut_down.get_lock():
        shut_down.value = 9
    return


if __name__ == '__main__':

    shut_down = multiprocessing.Value('i', 0)
    signal.signal(signal.SIGINT, signal_handler)

    io_queue_processor = IOQueueProcessor(reader, writer, dbo='childbo', output_file_name='outputfile.json')
    print(f"\nstarting {io_queue_processor.process_count} processes\n")
    io_queue_processor.start_up()

    cntr = 0
    for i in range(10):
        data = f"data{cntr}"
        print(f"sent {data}")
        cntr += 1
        io_queue_processor.process(data)
        if shut_down.value != 0:
            break

    print('\nfinishing up ...')
    io_queue_processor.finish_up()

    print(f"\n{cntr} records sent, {io_queue_processor.get_input_queue_read_cnt()} actually read, {io_queue_processor.get_output_queue_read_cnt()} actually written\n")




openc-load-childb.py
#! /usr/bin/env python3

import sys
import os
import glob
import sqlite3
import pandas
import time
import argparse

def make_database(dbname, child_file_types):
    child_dbo = sqlite3.connect(dbname, isolation_level=None, timeout=20)

    if 'address' in child_file_types:
        sql = 'CREATE TABLE IF NOT EXISTS address (' \
              ' company_number TEXT,' \
              ' jurisdiction_code TEXT,' \
              ' address_type TEXT,' \
              ' street_address TEXT,' \
              ' locality TEXT,' \
              ' region TEXT,' \
              ' postal_code TEXT,' \
              ' country TEXT,' \
              ' country_code TEXT,' \
              ' in_full TEXT,' \
              ' start_date TEXT,' \
              ' end_date TEXT' \
              ' )'
        child_dbo.cursor().execute(sql)

    if 'alias' in child_file_types:
        sql = 'CREATE TABLE IF NOT EXISTS alias (' \
              ' company_number TEXT,' \
              ' jurisdiction_code TEXT,' \
              ' name TEXT,' \
              ' type TEXT,' \
              ' start_date TEXT,' \
              ' end_date TEXT' \
              ' )'
        child_dbo.cursor().execute(sql)

    if 'identifier' in child_file_types:
        sql = 'CREATE TABLE IF NOT EXISTS identifier (' \
              ' company_number TEXT,' \
              ' jurisdiction_code TEXT,' \
              ' uid TEXT,' \
              ' identifier_system_code TEXT' \
              ' )'
        child_dbo.cursor().execute(sql)

    if 'telephone' in child_file_types:
        sql = 'CREATE TABLE IF NOT EXISTS telephone (' \
              ' company_number TEXT,' \
              ' jurisdiction_code TEXT,' \
              ' country_code TEXT,' \
              ' number TEXT,' \
              ' raw_number TEXT,' \
              ' number_type TEXT,' \
              ' start_date TEXT,' \
              ' end_date TEXT' \
              ' )'
        child_dbo.cursor().execute(sql)

    if 'website' in child_file_types:
        sql = 'CREATE TABLE IF NOT EXISTS website (' \
              ' company_number TEXT,' \
              ' jurisdiction_code TEXT,' \
              ' country_code TEXT,' \
              ' url TEXT,' \
              ' raw_url TEXT,' \
              ' number_type TEXT,' \
              ' start_date TEXT,' \
              ' end_date TEXT' \
              ' )'
        child_dbo.cursor().execute(sql)

    child_dbo.close()


def import_file(child_dbo, filedata, record_chunk_size):
    filetype = filedata[0]
    filename = filedata[1]
    print(f"{filename}: started")

    if os.path.splitext(filename)[1].upper() == '.GZ':
        file_compression = 'gzip'
    else:
        file_compression = None

    cnt = 0
    for chunk in pandas.read_csv(filename, chunksize=record_chunk_size, encoding='utf-8', dtype = str, compression=file_compression):
        timer_start = time.time()
        chunk.to_sql(filetype, child_dbo, index=False, method='multi', chunksize=1000, if_exists='append')
        cnt += len(chunk)
        print(f"{filename}: {cnt:,} records loaded, batch rate {round(time.time() - timer_start, 1)} seconds")

    return f"{filename} completed!"


def index_database(child_dbo, filetype):
    print(f"indexing {filetype} ...")
    timer_start = time.time()
    child_dbo.cursor().execute(f'create index ix_{filetype} on {filetype} (company_number, jurisdiction_code)')
    print(f"indexing {filetype} completed in {round(time.time() - timer_start, 1)} seconds")
    return f"{filetype} indexing complete"


def complete_database(child_dbo):
    child_dbo.cursor().execute('create table finished (dummy integer)')


if __name__ == "__main__":

    record_chunk_size = 1000000

    parser = argparse.ArgumentParser()
    parser.add_argument('-i', '--input_file_dir', dest='input_file_dir', help='the name of the open corporates csv file directory')
    parser.add_argument('-c', '--child_database_name', dest='child_database_name', help='the name of the database file to create')
    args = parser.parse_args()

    if not args.input_file_dir or not os.path.isdir(args.input_file_dir):
        print('\nPlease supply a valid input file directory on the command line\n')
        sys.exit(1)

    if not args.child_database_name:
        print('\nPlease supply a sqlite database file name on the command line\n')
        sys.exit(1)

    # note: telephones and websites are experimental and may not be present

    child_file_types = []
    child_files = []
    for file_name in sorted(glob.glob(args.input_file_dir + os.sep + '*')):
        base_name = os.path.basename(file_name)
        if 'non_reg_addresses' in base_name:
            file_type = 'address'
        elif 'alternative_names' in base_name:
            file_type = 'alias'
        elif 'additional_identifiers' in base_name:
            file_type = 'identifier'
        elif 'telephone' in base_name:
            file_type = 'telephone'
        elif 'website' in base_name:
            file_type = 'website'
        else:
            continue
        child_file_types.append(file_type)
        child_files.append([file_type,file_name])
    if not child_files:
        print('\nNo child files found!')
        print('Child file names must contain: "*non_reg_addresses*", "*alternative_names*" or "*additional_identifiers*".\n')
        sys.exit(1)

    if os.path.exists(args.child_database_name):
        os.remove(args.child_database_name)
        #print(f"\n{dbname} already exists, please remove it first if you want to continue\n")
        #sys.exit(1)

    proc_start = time.time()
    print(f"\n{len(child_files)} files to load\n")
    make_database(args.child_database_name, child_file_types)

    child_dbo = sqlite3.connect(args.child_database_name)
    child_dbo_cursor = child_dbo.cursor()
    child_dbo_cursor.execute('pragma synchronous = 0')
    child_dbo_cursor.execute('pragma journal_mode = off')
    child_dbo_cursor.execute('pragma temp_store = MEMORY')
    child_dbo_cursor.execute('pragma locking_mode = EXCLUSIVE')
    child_dbo_cursor.execute('pragma isolation_level = None')
    for filedata in child_files:
        print()
        import_file(child_dbo, filedata, record_chunk_size)
    for filetype in child_file_types:
        print()
        index_database(child_dbo, filetype)
    complete_database(child_dbo)

    child_dbo.close()

    print(f"\nProcess completed in {round((time.time() - proc_start) / 60, 1)} minutes\n")


openc-child-db.py
#! /usr/bin/env python3

import sys
import os
import glob
import sqlite3
import pandas
import time
import multiprocessing
import concurrent.futures
import argparse

def make_database(dbname, child_file_types):
    child_dbo = sqlite3.connect(dbname, isolation_level=None, timeout=20)

    if 'address' in child_file_types:
        sql = 'CREATE TABLE IF NOT EXISTS address (' \
              ' company_number TEXT,' \
              ' jurisdiction_code TEXT,' \
              ' address_type TEXT,' \
              ' street_address TEXT,' \
              ' locality TEXT,' \
              ' region TEXT,' \
              ' postal_code TEXT,' \
              ' country TEXT,' \
              ' country_code TEXT,' \
              ' in_full TEXT,' \
              ' start_date TEXT,' \
              ' end_date TEXT' \
              ' )'
        child_dbo.cursor().execute(sql)

    if 'alias' in child_file_types:
        sql = 'CREATE TABLE IF NOT EXISTS alias (' \
              ' company_number TEXT,' \
              ' jurisdiction_code TEXT,' \
              ' name TEXT,' \
              ' type TEXT,' \
              ' start_date TEXT,' \
              ' end_date TEXT' \
              ' )'
        child_dbo.cursor().execute(sql)

    if 'identifier' in child_file_types:
        sql = 'CREATE TABLE IF NOT EXISTS identifier (' \
              ' company_number TEXT,' \
              ' jurisdiction_code TEXT,' \
              ' uid TEXT,' \
              ' identifier_system_code TEXT' \
              ' )'
        child_dbo.cursor().execute(sql)

    if 'telephone' in child_file_types:
        sql = 'CREATE TABLE IF NOT EXISTS telephone (' \
              ' company_number TEXT,' \
              ' jurisdiction_code TEXT,' \
              ' country_code TEXT,' \
              ' number TEXT,' \
              ' raw_number TEXT,' \
              ' number_type TEXT,' \
              ' start_date TEXT,' \
              ' end_date TEXT' \
              ' )'
        child_dbo.cursor().execute(sql)

    if 'website' in child_file_types:
        sql = 'CREATE TABLE IF NOT EXISTS website (' \
              ' company_number TEXT,' \
              ' jurisdiction_code TEXT,' \
              ' country_code TEXT,' \
              ' url TEXT,' \
              ' raw_url TEXT,' \
              ' number_type TEXT,' \
              ' start_date TEXT,' \
              ' end_date TEXT' \
              ' )'
        child_dbo.cursor().execute(sql)

    child_dbo.close()

def import_file(dbname, filedata, record_chunk_size):
    filetype = filedata[0]
    filename = filedata[1]
    print(f"{filename}: started")
    child_dbo = sqlite3.connect(dbname, isolation_level=None, timeout=20)
    child_dbo.cursor().execute("PRAGMA journal_mode=wal")
    child_dbo.cursor().execute("PRAGMA synchronous=0")

    if os.path.splitext(filename)[1].upper() == '.GZ':
        file_compression = 'gzip'
    else:
        file_compression = None

    cnt = 0
    for chunk in pandas.read_csv(filename, chunksize=record_chunk_size, encoding='utf-8', dtype = str, compression=file_compression):
        timer_start = time.time()
        chunk.to_sql(filetype, child_dbo, index=False, method='multi', chunksize=1000, if_exists='append')
        cnt += len(chunk)
        print(f"{filename}: {cnt:,} records loaded, last chunk took {round(time.time() - timer_start, 1)} seconds")

    child_dbo.close()
    return f"{filename} completed!"

def import_file2(child_dbo, filedata, record_chunk_size):
    filetype = filedata[0]
    filename = filedata[1]
    print(f"{filename}: started")

    if os.path.splitext(filename)[1].upper() == '.GZ':
        file_compression = 'gzip'
    else:
        file_compression = None

    cnt = 0
    for chunk in pandas.read_csv(filename, chunksize=record_chunk_size, encoding='utf-8', dtype = str, compression=file_compression):
        timer_start = time.time()
        chunk.to_sql(filetype, child_dbo, index=False, method='multi', chunksize=1000, if_exists='append')
        cnt += len(chunk)
        print(f"{filename}: {cnt:,} records loaded, last chunk took {round(time.time() - timer_start, 1)} seconds")

    return f"{filename} completed!"


def index_database(dbname, filetype):
    print(f"indexing {filetype} ...")
    timer_start = time.time()
    child_dbo = sqlite3.connect(dbname, isolation_level=None, timeout=20)
    child_dbo.cursor().execute("PRAGMA journal_mode=wal")
    child_dbo.cursor().execute("PRAGMA synchronous=0")
    child_dbo.cursor().execute(f'create index ix_{filetype} on {filetype} (company_number, jurisdiction_code)')
    child_dbo.close()
    print(f"indexing {filetype} completed in {round(time.time() - timer_start, 1)} seconds")
    return f"{filetype} indexing complete"

def index_database2(child_dbo, filetype):
    print(f"indexing {filetype} ...")
    timer_start = time.time()
    child_dbo.cursor().execute(f'create index ix_{filetype} on {filetype} (company_number, jurisdiction_code)')
    print(f"indexing {filetype} completed in {round(time.time() - timer_start, 1)} seconds")
    return f"{filetype} indexing complete"

def complete_database(dbname):
    child_dbo = sqlite3.connect(dbname, isolation_level=None, timeout=20)
    child_dbo.cursor().execute('create table finished (dummy integer)')
    child_dbo.close()

def complete_database2(child_dbo):
    child_dbo.cursor().execute('create table finished (dummy integer)')

if __name__ == "__main__":

    record_chunk_size = 1000000

    parser = argparse.ArgumentParser()
    parser.add_argument('-i', '--input_file_dir', dest='input_file_dir', help='the name of the open corporates csv file directory')
    parser.add_argument('-c', '--child_database_name', dest='child_database_name', help='the name of the database file to create')
    args = parser.parse_args()

    if not args.input_file_dir or not os.path.isdir(args.input_file_dir):
        print('\nPlease supply a valid input file directory on the command line\n')
        sys.exit(1)

    if not args.child_database_name:
        print('\nPlease supply a sqlite database file name on the command line\n')
        sys.exit(1)

    # note: telephones and websites are experimental and may not be present

    child_file_types = []
    child_files = []
    for file_name in sorted(glob.glob(args.input_file_dir + os.sep + '*')):
        base_name = os.path.basename(file_name)
        if 'non_reg_addresses' in base_name:
            file_type = 'address'
        elif 'alternative_names' in base_name:
            file_type = 'alias'
        elif 'additional_identifiers' in base_name:
            file_type = 'identifier'
        elif 'telephone' in base_name:
            file_type = 'telephone'
        elif 'website' in base_name:
            file_type = 'website'
        else:
            continue
        child_file_types.append(file_type)
        child_files.append([file_type,file_name])
    if not child_files:
        print('\nNo child files found!')
        print('Child file names must contain: "*non_reg_addresses*", "*alternative_names*" or "*additional_identifiers*".\n')
        sys.exit(1)

    if os.path.exists(args.child_database_name):
        os.remove(args.child_database_name)
        #print(f"\n{dbname} already exists, please remove it first if you want to continue\n")
        #sys.exit(1)
    make_database(args.child_database_name, child_file_types)

    single_threaded = True


    proc_start = time.time()

    print(f"\n{len(child_files)} files to load\n")

    # load
    if single_threaded:
        child_dbo = sqlite3.connect(args.child_database_name)
        child_dbo_cursor = child_dbo.cursor()
        child_dbo_cursor.execute('pragma synchronous = 0')
        child_dbo_cursor.execute('pragma cache_size = -16000000')
        child_dbo_cursor.execute('pragma secure_delete = 0')
        child_dbo_cursor.execute('pragma page_size = 65536')
        child_dbo_cursor.execute('pragma journal_mode = off')
        child_dbo_cursor.execute('pragma temp_store = MEMORY')
        child_dbo_cursor.execute('pragma locking_mode = EXCLUSIVE')
        child_dbo_cursor.execute('pragma isolation_level = None')
        for filedata in child_files:
            import_file2(child_dbo, filedata, record_chunk_size)
        for filetype in child_file_types:
            index_database2(child_dbo, filetype)
        complete_database2(child_dbo)

    else:
        with concurrent.futures.ProcessPoolExecutor() as executor:
            futures = {
                executor.submit(import_file, args.child_database_name, filedata, record_chunk_size):
                filedata for filedata in child_files
            }
            while futures:
                done, futures = concurrent.futures.wait(futures, return_when=concurrent.futures.FIRST_COMPLETED)
                for fut in done:
                    print(fut.result())

        # index
        with concurrent.futures.ProcessPoolExecutor() as executor:
            futures = {
                executor.submit(index_database, args.child_database_name, filetype):
                filetype for filetype in child_file_types
            }
            while futures:
                done, futures = concurrent.futures.wait(futures, return_when=concurrent.futures.FIRST_COMPLETED)
                for fut in done:
                    print(fut.result())

        complete_database(args.child_database_name)

    print(f"\nProcess completed in {round((time.time() - proc_start) / 60, 1)} minutes\n")
